Implement PSA protected storage & restructure PSA storage implementation

davidsaada opened the pull request:
### Description
This PR's main purpose is the implementation of the PSA Protected Storage (PS) feature. As this implementation is almost identical to the one we already have in PSA's Internal Trusted Storage (ITS), it triggered a deeper change in the entire implementation of PSA storage.

So, changes included here are:
- Move all PSA storage code under psa/storage directory
- Create a global PSA error codes header, eliminating ITS specific ones (backed by [a change in the PSA spec](https://github.com/ARMmbed/psa_trusted_storage_api/pull/49))
- Create a common header file for PSA storage type definitions, eliminating ITS specific ones (backed by the same change in the PSA spec)
- Create a common implementation for PS & ITS
- Implement protected storage feature
- Change ITS test to be common to PS as well
- Modify affected PSA crypto code (temporary, for CI sake only)

**Important notes:**
~- Changes in the mbed-crypto code, included in this PR, are temporary. Changes should eventually be taken from [PR #55 in the mbed-crypto repo](https://github.com/ARMmbed/mbed-crypto/pull/55). They are included here for CI sake only. After both PRs are approved, they should be merged together.~
- This PR is based on the current PSA spec. There's an ongoing [discussion](https://github.com/ARMmbed/psa_trusted_storage_api/issues/53) regarding the way PSA get functions should behave that's still open. As we strive to push this PR for 5.12, we can't wait for this discussion to be resolved, but if it's resolved in a reasonable time frame, PR may change accordingly.

Tested with all PSA tests on K64F and FUTURE_SEQUANA boards.

### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Release Notes
- Restructure PSA storage implementation (directory structure and common code)
- Add the PSA protected storage feature

Commit:
Implement PSA protected storage & restructure PSA storage implementation

- Move all PSA storage code under psa/storage directory
- Create a global PSA error codes header, eliminating ITS specific ones
- Create a common header file for PSA storage type definitions,
  eliminating ITS specific ones
- Create a common implementation for PS & ITS
- Implement protected storage feature
- Change ITS test to be common to PS as well

ciarmcom commented:
@davidsaada, thank you for your changes.
@ARMmbed/mbed-os-tls @ARMmbed/mbed-os-crypto @ARMmbed/mbed-os-maintainers please review.

simonbutcher commented:
This has no impact on @ARMmbed/mbed-os-tls so please remove as approvers.

0xc0170 commented:
@davidsaada This will need release notes section https://os.mbed.com/docs/mbed-os/v5.11/contributing/workflow.html#functionality-change

davidsaada commented:
This depends on #9702 now in order to fix travis issues.

mikisch81 commented:
@davidsaada Can you look at #9653 and see what is the impact of the changes there on this PR (or the opposite, whichever is merged first)

cmonr commented:
> This depends on #9702 now in order to fix travis issues.

@davidsaada Travis issues?

davidsaada commented:
> 
> 
> > This depends on #9702 now in order to fix travis issues.
> 
> @davidsaada Travis issues?

Should have been more specific: It failed the psa-autogen travis check before depending on that PR.

davidsaada commented:
> 
> 
> @davidsaada Can you look at #9653 and see what is the impact of the changes there on this PR (or the opposite, whichever is merged first)

@mikisch81 The PRs are almost orthogonal. They only collide in `its_init`, where you added `get_its_kvstore_instance`, which can easily be merged (in either direction).

davidsaada commented:
Rebased, no longer depends on #9702 .
In addition, split the change into two commits, where mbed-crypto files are separated, so they can easily be removed from change once brought from mbed-crypto repo.

Patater committed:
crypto: test: Use new PSA error codes

The PSA Framework and PSA Crypto error code names have been unified and
renamed in some cases. Use the new unified names.

Patater committed:
crypto: Update Mbed Crypto to 1.0.0d3

0xc0170 commented on components/TARGET_PSA/services/storage/ps/COMPONENT_NSPE/protected_storage.cpp:
```diff
@@ -0,0 +1,125 @@
+/* Copyright (c) 2018 ARM Limited
```
2019 for new files

davidsaada replied:
Fixed.

0xc0170 commented on components/TARGET_PSA/inc/psa/error.h:
```diff
@@ -0,0 +1,38 @@
+/* psa/error.h
+Standard error codes for the SPM and RoT Services
```
this file needs license header file

davidsaada replied:
Fixed.

0xc0170 commented on TESTS/psa/its_ps/main.cpp:
```diff
@@ -0,0 +1,198 @@
+/*
+* Copyright (c) 2018 ARM Limited. All rights reserved.
```
2019 year

davidsaada replied:
Fixed (though this one was moved).

0xc0170 commented on components/TARGET_PSA/services/storage/common/psa_storage_common_impl.h:
```diff
@@ -0,0 +1,50 @@
+/* Copyright (c) 2018 ARM Limited
```
2019 also here

davidsaada replied:
Fixed.

0xc0170 commented on components/TARGET_PSA/services/storage/its/COMPONENT_PSA_SRV_IMPL/pits_impl.cpp:
```diff
@@ -0,0 +1,123 @@
+/* Copyright (c) 2018 ARM Limited
```
2019

davidsaada replied:
Fixed (though this one was moved).

0xc0170 commented on components/TARGET_PSA/services/storage/common/psa_storage_common_impl.h:
```diff
@@ -0,0 +1,50 @@
+
+typedef psa_status_t (*migrate_func_t)(mbed::KVStore *kvstore, const psa_storage_version_t *old_version, const psa_storage_version_t *new_version);
+
+void psa_storage_handle_version(mbed::KVStore *kvstore, const char *version_key, const psa_storage_version_t *version,
```
no documentation for these functions? _impl.h are internal headers?

davidsaada replied:
Yes, they are. They are the worker functions for the PS & ITS features.

0xc0170 reviewed (approved):
Just one questions regarding _impl headers - lacking the docs there

Patater committed:
crypto: Update to Mbed Crypto 1.0.0d4

Commit:
Update crypto IPC handling of negative errors

Commit:
Add flash IAP to sequana_psa m4

Patater reviewed (approved):
Crypto files LGTM. I re-ran the importer and verified no changes, so this is an accurate version of Mbed Crypto 1.0.0d4.
